chain/chunks/src/lib.rs

```diff
@@ -677,12 +704,37 @@ impl ShardsManager {
 
         let can_reconstruct = entry.parts.len() >= self.runtime_adapter.num_data_parts();
 
-        if have_all_parts {
+        let chunk_producer = self.runtime_adapter.get_chunk_producer(
+            &self.runtime_adapter.get_epoch_id_from_prev_block(&prev_block_hash)?,
+            header.inner.height_created,
+            header.inner.shard_id,
+        )?;
+        self.track_seals();
+
+        if have_all_parts && !self.resented_by.contains(&chunk_producer) {
             self.encoded_chunks.insert_chunk_header(partial_encoded_chunk.shard_id, header.clone());
         }
         let entry = self.encoded_chunks.get(&chunk_hash).unwrap();
 
-        if have_all_parts && have_all_receipts {
+        // TODO need help to refactor
+        let seal = if let Some(seal) = self.seals.get(&chunk_hash) {
+            seal
+        } else {
+            let seal = self.produce_seal(
+                &prev_block_hash,
+                header.inner.height_created,
+                header.inner.shard_id,
+            )?;
+            self.seals.insert(chunk_hash.clone(), seal);
+            self.seals.get(&chunk_hash).unwrap()
+        };
+        let (have_all_seal, part_ords_to_remove) = self.process_seal(seal, entry)?;
```

Comment:
`process_seal` doesn't use `self`, so remove `self` there, this will allow you go `get_mut` the seal above, pass mutable `seal` into `process_seal`, and make `process_seal` actually remove the parts, not return them.